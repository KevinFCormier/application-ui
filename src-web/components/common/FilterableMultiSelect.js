/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { MultiSelect } from 'carbon-components-react'
import msgs from '../../../nls/platform.properties'

class FilterableMultiSelect extends React.Component {

  static propTypes = {
    activeFilters: PropTypes.array,
    availableFilters: PropTypes.array,
    failure: PropTypes.bool,
    fetching: PropTypes.bool,
    filterType: PropTypes.string,
    onChange: PropTypes.func,
    title: PropTypes.string
  }
  /**
   * The <MultiSelect> component uses a shallow compare when computing which items
   * are selected. This function helps to make sure that the array of selected items
   * reference the objects from the items array for the shallow compare to work.
   */
  getSelectedFilters = (items = [], selected = []) =>
    items.filter(i => selected.find(f => i.label === f.label))


  handleFilter = selection => {
    this.props.onChange(this.props.filterType, selection.selectedItems)
    this.updateTooltip()
  }

  componentDidMount () {
    this.updateTooltip()
  }

  updateTooltip () {
    this.multiSelect.inputNode.title = this.tooltip.join('\n')
  }

  render() {
    const {
      title,
      availableFilters,
      activeFilters=[],
      fetching,
      failure} = this.props

    this.tooltip = []
    if (failure) {
      this.tooltip.push(msgs.get('resource.error', this.context.locale))
    } else if (fetching) {
      this.tooltip.push(msgs.get('resource.loading', this.context.locale))
    } else if (!activeFilters.length) {
      this.tooltip.push(msgs.get('resource.filterAll', this.context.locale))
    } else {
      this.tooltip = activeFilters.map(filter => filter.label).sort()
    }
    return (
      <div className='multi-select-filter'>
        <div className='multi-select-filter-title'>
          {title}
        </div>
        <MultiSelect.Filterable
          ref={this.saveMultiSelectRef}
          placeholder={this.tooltip.join(', ')}
          items={availableFilters}
          initialSelectedItems={this.getSelectedFilters(availableFilters, activeFilters)}
          onChange={this.handleFilter}
          disabled={fetching}
        />
      </div>
    )
  }

  saveMultiSelectRef = ref => {
    this.multiSelect = ref
  }
}

export default FilterableMultiSelect
