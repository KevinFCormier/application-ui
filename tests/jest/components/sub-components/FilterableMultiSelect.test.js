/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { shallow } from 'enzyme'

import FilterableMultiSelect from '../../../../src-web/components/common/FilterableMultiSelect'

const mockData = {
  title: 'test',
  type: 'test',
  fetching: false,
  failure: false,
  availableFilters: [

  ],
  activeFilters: [

  ],
  onSelectedFilterChange: jest.fn()
}

describe('FilterableMultiSelect components', () => {

  it('renders as expected', () => {
    const component = shallow(
      <FilterableMultiSelect
        filterType={mockData.type}
        title={mockData.title}
        availableFilters={mockData.availableFilters}
        activeFilters={mockData.activeFilters}
        onChange={mockData.onSelectedFilterChange}
        fetching={mockData.fetching}
        failure={mockData.failure}
      />
    )
    expect(component).toMatchSnapshot()
  })

})
