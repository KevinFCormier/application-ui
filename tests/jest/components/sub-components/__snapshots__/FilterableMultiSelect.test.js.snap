// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableMultiSelect components renders as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterableMultiSelect
    activeFilters={Array []}
    availableFilters={Array []}
    failure={false}
    fetching={false}
    filterType="test"
    onChange={[MockFunction]}
    title="test"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="multi-select-filter-title"
        >
          test
        </div>,
        <FilterableMultiSelect
          compareItems={[Function]}
          disabled={false}
          filterItems={[Function]}
          initialSelectedItems={Array []}
          itemToString={[Function]}
          items={Array []}
          locale="en"
          onChange={[Function]}
          placeholder="all"
          sortItems={[Function]}
        />,
      ],
      "className": "multi-select-filter",
      "id": "test",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "test",
          "className": "multi-select-filter-title",
        },
        "ref": null,
        "rendered": "test",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "compareItems": [Function],
          "disabled": false,
          "filterItems": [Function],
          "initialSelectedItems": Array [],
          "itemToString": [Function],
          "items": Array [],
          "locale": "en",
          "onChange": [Function],
          "placeholder": "all",
          "sortItems": [Function],
        },
        "ref": [Function],
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="multi-select-filter-title"
          >
            test
          </div>,
          <FilterableMultiSelect
            compareItems={[Function]}
            disabled={false}
            filterItems={[Function]}
            initialSelectedItems={Array []}
            itemToString={[Function]}
            items={Array []}
            locale="en"
            onChange={[Function]}
            placeholder="all"
            sortItems={[Function]}
          />,
        ],
        "className": "multi-select-filter",
        "id": "test",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "test",
            "className": "multi-select-filter-title",
          },
          "ref": null,
          "rendered": "test",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "compareItems": [Function],
            "disabled": false,
            "filterItems": [Function],
            "initialSelectedItems": Array [],
            "itemToString": [Function],
            "items": Array [],
            "locale": "en",
            "onChange": [Function],
            "placeholder": "all",
            "sortItems": [Function],
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
